pkg/controller/deployment/deployment_controller.go

```diff
@@ -599,6 +605,24 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
 	return true, err
 }
 
+func (dc *DeploymentController) cleanupOldRcs(controllers []*api.ReplicationController, deployment extensions.Deployment) error {
+	diff := len(controllers) - *deployment.Spec.RevisionHistoryLimit
+	if diff <= 0 {
+		return nil
+	}
+
+	sort.Sort(byCreationTimestamp(controllers))
+	// TODO: This should be parallelized.
+	for i := 0; i < diff; i++ {
+		controller := controllers[i]
+		if err := dc.client.ReplicationControllers(controller.Namespace).Delete(controller.Name); err != nil {
```

Comment:
This needs to not delete RCs with non-zero replica counts. I think it's fine to count all but the latest RC against the limit, and to start deleting oldest first, but we should skip those with non-zero counts.

Scenario: For a Deployment with RevisionHistoryLimit=10, a user makes a dozen changes in rapid succession, generating a dozen new RCs, each of which scales up a few pods before the next RC is created.